recipe-client-addon/lib/StudyStorage.jsm

```diff
@@ -0,0 +1,108 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * @typedef {Object} Study
+ * @property {string} name
+ *   Unique name of the study
+ * @property {number} version
+ *   Study add-on version number
+ * @property {string} description
+ *   Description of the study and its intent.
+ * @property {string} studyStartDate
+ *   ISO-formatted date string of when the study was started.
+ * @property {string} studyEndDate
+ *   ISO-formatted date string of when the study was ended.
+ * @property {string} addonId
+ *   Add-on ID for this particular study.
+ * @property {boolean} active
+ *   Is the study still running?
+ */
+
+"use strict";
+
+const {utils: Cu} = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "IndexedDB", "resource://gre/modules/IndexedDB.jsm");
+
+this.EXPORTED_SYMBOLS = ["StudyStorage"];
+
+const DB_OPTIONS = {
+  namespace: "shield-study-addons",
+  version: 1,
+  storage: "persistent",
+  schema: {
+    name: { autoIncrement: false, unique: true },
+    version: { autoIncrement: false, unique: false },
+    description: { autoIncrement: false, unique: false },
+    studyStartDate: { autoIncrement: false, unique: false },
+    studyEndDate: { autoIncrement: false, unique: false },
+    addonId: { autoIncrement: false, unique: false },
+    active: { autoIncrement: false, unique: false },
+  },
+};
+
+const dbName = "shield-studies";
+let database;
+
+async function getDatabase() {
+  if (!database) {
+    database = await IndexedDB.open(dbName, DB_OPTIONS, db => {
+      db.createObjectStore(dbName, { keyPath: "name" });
+    });
+  }
+  return database;
+}
+
+this.StudyStorage = {
+  async clearAllData() {
+    if (database) {
+      await database.objectStore(dbName, "readwrite").clear();
+    }
+  },
+
+  async closeConnection() {
+    if (database) {
+      await database.close();
+      database = null;
+    }
+  },
+
+  async getStudy(studyName) {
+    const db = await getDatabase();
+
+    return db.objectStore(dbName, "readonly").get(studyName);
+  },
+
+  async saveStudy(study) {
```

Comment:
I remember looking for a schema library in mozilla-central at least one time before this, plus we probably eventually want to shift to a single storage solution across the entire add-on, instead of the split between IndexedDB and JSONFile that we'll have once this lands. I think we'll be doing enough schema stuff that pulling it in is worth it, although ideally mozilla-central would eventually ship it built-in and we wouldn't have to pull it in ourselves.